### ISSUE TITLE
Establish consistent pattern for testing exact errors

### ISSUE THREAD

**Issue by grjte** [maticnetwork/miden#43]
As discussed in PR #40 , we want tests to be checking for exact errors according to Rust best practices.

TL;DR: there are different options and no perfect consensus, so we should make an opinionated decision

## Background

It seems like Rust actually doesn't handle error testing as ergonomically as most other things, so there isn't an opinionated "right answer." Most errors in Rust implement the PartialEq trait, which allows using assert_eq! in tests to check exact errors. However, a few kinds of errors don't, most notably std::io::Error. See the reference discussions below for more detailed reasoning, but it can be summarized by "It's an unfortunate ergonomic loss that io::Result<T> instances can't be equated, but indicating that io::Error has an implementation of PartialEq is somewhat of a lie because the only type of error that can be equal is two OS errors"

### Option 1: Implement PartialEq trait for custom errors
The info above suggests that either deriving PartialEq (where possible) or manually implementing it for custom error types and then using ```assert_eq!``` is an acceptable idiomatic option (as recommended in the referenced Stack Overflow discussion as well).

### Option 2: Use ```should_panic``` or return a ```Result<T,E>``` from the test
The [book](https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic) presents 2 options:
1. use ```should_panic``` with an expectation string
2. give the test a Result return type and return the expected Error

This is possibly the "best practice" given that it's in the book, but it's not always fine-grained enough & many people seem dissatisfied with it. Additionally, both options mean that there will be strictly one error checked per unit test, which makes sense with the idea of a "unit test" but could cause a blowup in the number of tests, which may or may not be desirable (needs an opinionated decision).

```should_panic``` can be improved upon by using ```catch_unwind``` which "allows you to put the code you expect to panic into a closure, and only panics emitted by that code will be considered expected (i.e. a passing test)." See the second SO reference for a discussion & links to examples of this approach being used in the Rust standard library. This approach cannot catch non-unwinding panics.

Final note: using ```should_panic``` can give different output depending on how it's executed. "As you can see, the output of the second should_panic test case contains a backtrace. However, if you run your tests with the simple cargo test <test_name> command, the output of the both test cases are the same. So, be careful when you configure your CI/CD system!" (see the reference from Yury Zhauniarovich)

### Option 3: Use [```assert_matches```](https://crates.io/crates/assert_matches) crate
We could use this crate to test pattern matching. It gives more control than the second option without the overhead of implementing PartialEq. This was mentioned in the Rust discussions referenced below, but I haven't seen it mentioned otherwise.

## References:
unit testing errors (PartialEq):
https://stackoverflow.com/questions/57234140/how-to-assert-io-errors-in-rust
testing with should_panic (catch_unwind): 
https://stackoverflow.com/questions/26469715/how-do-i-write-a-rust-unit-test-that-ensures-that-a-panic-has-occurred
article on ins & outs of testing errors: 
https://zhauniarovich.com/post/2021/2021-01-testing-errors-in-rust/

### Discussions with the Rust team:
https://github.com/rust-lang/rust/pull/34192
https://github.com/rust-lang/rust/issues/34158

**Comment by bobbinth** [comment]
Thank you for such an exhaustive description! I'm thinking we can go with the first option since we don't rely `std::io::Error` and adding `PartialEq` to all errors shouldn't be too big of an issue.

**Comment by grjte** [comment]
Sounds good to me!
